brainstorming.md

```diff
@@ -0,0 +1,372 @@
+
+> ⚠️ **Issue** `raw` -- do we need `raw.X` and `raw.var` at all, or can we model them with existing structures? **Proposal**: the current `X` and `varm` can represent typical `raw` semantics (filtered subset, or transformed values). Assume that higher level usage conventions will utilize these to represent raw, eg, `X['raw']` and `varm['raw']`, or alternatively by adding additional columns to `var` that indicate filtering status.
+
+> ⚠️ **Issue** `obsm` and `varm` schema -- if these are RecordBatches indexed by obs_id / var_id, then they are essentially identical to obs/var slots. Are they necessary, or can we represent obsm/varm with conventions layered on top of obs/var? if they are different, how? The only current difference is that `obs` and `var` are required to have an entry for each value on that dimension, whereas `obsm`/`varm` do not require all dimension IDs to exist (ie, they are sparse like X). **Proposal**: retain, but differentiate by considering obsm/varm as secondary annotations on the axis, which do **not** require a value for each obs_id/var_id (ie are **sparse**).
+
+> ⚠️ **Issue** `obsm` and `varm` names - field inspiration and name were borrowed from AnnData, but are misleading in this context as both are essentially a collection of SOMADataFrame values, not a collection of dense "matrix" in the AnnData sense. We should consider renaming them. While they can express the typical use case of AnnData.obsm/varm, they are better thought of as a collection of DataFrame containing secondary annotations for the obs/var axis. Possible ideas for better names: obsd/vard ('d' as in dataframe), obs_axis/var_axis, "axis annotation layers", or?
+
+# Functional Operations
+
+The SOMA API includes functional capabilities around the [SOMA data model](#datamodel.md). The specifics of how these operations manifest in any given language and computing environment is defined in per-language specifications (**to be created**), but each implementation will minimally support the functional operations defined here. For example, it is likely that a Python implementation will prefer `__getattr__` over an explicit `get()` method, and will add other Pythonic functionality to facilitate use.
+
+As any given storage "engine" upon which SOMA is implemented may have additional features and capabilities required to support advanced use cases, it is expected that SOMA implementations will expose engine-specific features. Where possible, these should be implemented to avoid conflict with future changes to the common SOMA API.
+
+> ⚠️ **Issue** - this spec needs to provide guidance on HOW to do that. Fro example, can we carve out some namespace for extensions in the functional API?
+
+In the following, a pseudo-object-oriented syntax is used, eg, `obj.op()` is calling `op()` upon `obj`. Where a type is used, eg, SOMA.op(), it is a `static` operation (ie, no state other than parameters provided).
```

Comment:
good call -- various readers with strengths/weaknesses in various languages will be reading this & this is very welcome-mat-come-on-in :)